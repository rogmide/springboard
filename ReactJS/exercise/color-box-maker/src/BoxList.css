.BoxList {
  margin: 10px;
}
