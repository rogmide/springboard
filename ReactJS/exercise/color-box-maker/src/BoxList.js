import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import "./BoxList.css";
import { v4 as uuidv4 } from "uuid";

const BoxList = () => {
  const INITIAL_STATE = [
    // default box to work with app
    { id: uuidv4(), width: "100", height: "100", backgroundColor: "green" },
  ];

  const [box, setBox] = useState(INITIAL_STATE);

  // Add box to the Array
  const addBox = (newBox) => {
    setBox((box) => [...box, { ...newBox, id: uuidv4() }]);
  };

  // Remove box from the Array
  const removeBox = (id) => {
    setBox(box.filter((b) => b.id !== id));
  };

  return (
    <>
      {" "}
      <div className="BoxList">
        <h1>Box List</h1>
        {box.map(({ id, width, height, backgroundColor }) => (
          <Box
            removeBox={removeBox}
            id={id}
            width={width}
            height={height}
            backgroundColor={backgroundColor}
            key={id}
          />
        ))}
      </div>
      <NewBoxForm addBox={addBox} />
    </>
  );
};

export default BoxList;
