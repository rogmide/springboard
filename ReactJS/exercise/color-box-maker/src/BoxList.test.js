import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("render without crashing", function () {
  render(<BoxList />);
});

it("should match snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

// try get by test ID test-box
it("should add a new box", () => {
  const { queryByText, getByLabelText } = render(<BoxList />);
  const inputWidth = getByLabelText("Width:");
  const inputHeight = getByLabelText("Height:");
  const inputBackColor = getByLabelText("Background Color:");
  const btn = queryByText("Add Box");

  const btn1 = queryByText("X");
  expect(queryByText("X")).toBeInTheDocument();
  expect(queryByText("X")).toHaveStyle({ backgroundColor: "green" });

  fireEvent.click(btn1);
  expect(queryByText("X")).not.toBeInTheDocument();

  // mocking a input on the app passing a obj
  fireEvent.change(inputWidth, { target: { value: "100" } });
  fireEvent.change(inputHeight, { target: { value: "100" } });
  fireEvent.change(inputBackColor, { target: { value: "red" } });
  fireEvent.click(btn);
  expect(queryByText("X")).toBeInTheDocument();
  //   expect(queryByText("X")).toHaveStyle("background-color: red");
});
