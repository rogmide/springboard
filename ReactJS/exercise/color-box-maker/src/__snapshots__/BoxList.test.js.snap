// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="BoxList"
  >
    <h1>
      Box List
    </h1>
    <div
      class="Box"
      style="width: 100px; height: 100px; background-color: green;"
    >
      <h3>
        X
      </h3>
    </div>
  </div>
  <h1>
    Box Form
  </h1>
  <form
    class="NewBoxForm"
  >
    <label
      for="width"
    >
      Width: 
    </label>
    <input
      id="width"
      max="500"
      min="20"
      name="width"
      placeholder="width"
      type="number"
      value=""
    />
    <label
      for="height"
    >
      Height: 
    </label>
    <input
      id="height"
      max="500"
      min="20"
      name="height"
      placeholder="height"
      type="number"
      value=""
    />
    <label
      for="backgroundColor"
    >
      Background Color: 
    </label>
    <input
      id="backgroundColor"
      name="backgroundColor"
      placeholder="backgroundColor"
      type="text"
      value=""
    />
    <br />
    <button>
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
